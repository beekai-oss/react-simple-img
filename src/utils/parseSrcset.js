export default function parseSrcset(s) {
  const sSources = s.match(/\s*(?:[\S]*)(?:\s+(?:-?(?:\d+(?:\.(?:\d+)?)?|\.\d+)[a-zA-Z]))?(?:\s*,)?/g);
  const sources = [];

  for (let i = 0; i < sSources.length; i++) {
    let sSource = sSources[i].trim();

    if (sSource.substr(-1) === ',') {
      sSource = sSource.substr(0, sSource.length - 1).trim();
    }

    const parts = sSource.split(/\s+/, 2);

    if (parts.length === 0 || (parts.length === 1 && !parts[0]) || (parts.length === 2 && !parts[0] && !parts[1])) {
      continue;
    }

    const url = parts[0];

    if (parts.length === 1 || (parts.length === 2 && !parts[1])) {
      // If no "w" or "x" specified, we assume it's "1x".
      sources.push({ url, width: undefined, dpr: 1 });
    } else {
      const spec = parts[1].toLowerCase();
      const lastChar = spec.substring(spec.length - 1);
      if (lastChar === 'w') {
        sources.push({ url, width: parseFloat(spec), dpr: undefined });
      } else if (lastChar === 'x') {
        sources.push({ url, width: undefined, dpr: parseFloat(spec) });
      }
    }
  }

  return sources;
}
