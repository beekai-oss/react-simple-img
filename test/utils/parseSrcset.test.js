import parseSrcset from '../../src/utils/parseSrcset';

describe('parseSrcset', () => {
  it('should parse src set correctly', () => {
    expect(parseSrcset('medium.jpg 1000w, large.jpg 2000w')).toMatchSnapshot();
    expect(parseSrcset('large.jpg 3x, medium.jpg 2x, large.jpg')).toMatchSnapshot();
  });

  it('should throw error when invalid srcset is passed', () => {
    try {
      parseSrcset(23192381293123);
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });
});
