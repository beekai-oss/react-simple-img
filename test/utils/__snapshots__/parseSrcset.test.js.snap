// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseSrcset should parse src set correctly 1`] = `
Array [
  Object {
    "dpr": undefined,
    "url": "medium.jpg",
    "width": 1000,
  },
  Object {
    "dpr": undefined,
    "url": "large.jpg",
    "width": 2000,
  },
]
`;

exports[`parseSrcset should parse src set correctly 2`] = `
Array [
  Object {
    "dpr": 3,
    "url": "large.jpg",
    "width": undefined,
  },
  Object {
    "dpr": 2,
    "url": "medium.jpg",
    "width": undefined,
  },
  Object {
    "dpr": 1,
    "url": "large.jpg",
    "width": undefined,
  },
]
`;

exports[`parseSrcset should throw error when invalid srcset is passed 1`] = `[TypeError: srcset.match is not a function]`;
